## Get the data 

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('data_insy695.csv')
df.head()

Customer       State  Customer Lifetime Value Response  Coverage Education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  Effective To Date EmploymentStatus Gender  Income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4            2/3/11         Employed      M   43836  ...   

  Months Since Policy Inception Number of Open Complaints  Number of Policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      Policy Type        Policy  Renew Offer Type  Sales Channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  Total Claim Amount  Vehicle Class Vehicle Size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

## Observations

In [3]:
df.columns

Index(['Customer', 'State', 'Customer Lifetime Value', 'Response', 'Coverage',
       'Education', 'Effective To Date', 'EmploymentStatus', 'Gender',
       'Income', 'Location Code', 'Marital Status', 'Monthly Premium Auto',
       'Months Since Last Claim', 'Months Since Policy Inception',
       'Number of Open Complaints', 'Number of Policies', 'Policy Type',
       'Policy', 'Renew Offer Type', 'Sales Channel', 'Total Claim Amount',
       'Vehicle Class', 'Vehicle Size'],
      dtype='object')

In [4]:
df.isna().sum() #No NA values 

Customer                         0
State                            0
Customer Lifetime Value          0
Response                         0
Coverage                         0
Education                        0
Effective To Date                0
EmploymentStatus                 0
Gender                           0
Income                           0
Location Code                    0
Marital Status                   0
Monthly Premium Auto             0
Months Since Last Claim          0
Months Since Policy Inception    0
Number of Open Complaints        0
Number of Policies               0
Policy Type                      0
Policy                           0
Renew Offer Type                 0
Sales Channel                    0
Total Claim Amount               0
Vehicle Class                    0
Vehicle Size                     0
dtype: int64

In [5]:
df.info() #we check the datatype of variables in the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9134 entries, 0 to 9133
Data columns (total 24 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Customer                       9134 non-null   object 
 1   State                          9134 non-null   object 
 2   Customer Lifetime Value        9134 non-null   float64
 3   Response                       9134 non-null   object 
 4   Coverage                       9134 non-null   object 
 5   Education                      9134 non-null   object 
 6   Effective To Date              9134 non-null   object 
 7   EmploymentStatus               9134 non-null   object 
 8   Gender                         9134 non-null   object 
 9   Income                         9134 non-null   int64  
 10  Location Code                  9134 non-null   object 
 11  Marital Status                 9134 non-null   object 
 12  Monthly Premium Auto           9134 non-null   i

In [6]:
#Looking at categorical data to see their distribution in terms of count per class: 

In [7]:
df['State'].value_counts() 

California    3150
Oregon        2601
Arizona       1703
Nevada         882
Washington     798
Name: State, dtype: int64

In [8]:
df['Response'].value_counts()

No     7826
Yes    1308
Name: Response, dtype: int64

In [8]:
df['Coverage'].value_counts()

Basic       5568
Extended    2742
Premium      824
Name: Coverage, dtype: int64

In [9]:
df['Education'].value_counts()

Bachelor                2748
College                 2681
High School or Below    2622
Master                   741
Doctor                   342
Name: Education, dtype: int64

In [10]:
df['Policy'].value_counts()

Personal L3     3426
Personal L2     2122
Personal L1     1240
Corporate L3    1014
Corporate L2     595
Corporate L1     359
Special L2       164
Special L3       148
Special L1        66
Name: Policy, dtype: int64

In [11]:
df['Policy Type'].value_counts()

Personal Auto     6788
Corporate Auto    1968
Special Auto       378
Name: Policy Type, dtype: int64

In [39]:
df.describe() #We can see that the fetaures are in different scales 

Customer Lifetime Value        Income  Monthly Premium Auto  \
count              9134.000000   9134.000000           9134.000000   
mean               8004.940475  37657.380009             93.219291   
std                6870.967608  30379.904734             34.407967   
min                1898.007675      0.000000             61.000000   
25%                3994.251794      0.000000             68.000000   
50%                5780.182197  33889.500000             83.000000   
75%                8962.167041  62320.000000            109.000000   
max               83325.381190  99981.000000            298.000000   

       Months Since Last Claim  Months Since Policy Inception  \
count              9134.000000                    9134.000000   
mean                 15.097000                      48.064594   
std                  10.073257                      27.905991   
min                   0.000000                       0.000000   
25%                   6.000000                      24.000000   
50%                  14.000000                      48.000000   
75%                  23.000000                      71.000000   
max                  35.000000                      99.000000   

       Number of Open Complaints  Number of Policies  Total Claim Amount  
count                9134.000000         9134.000000         9134.000000  
mean                    0.384388            2.966170          434.088794  
std                     0.910384            2.390182          290.500092  
min                     0.000000            1.000000            0.099007  
25%                     0.000000            1.000000          272.258244  
50%                     0.000000            2.000000          383.945434  
75%                     0.000000            4.000000          547.514839  
max                     5.000000            9.000000         2893.239678

In [38]:
df.corr() #Correlations between categorical variables 

Customer Lifetime Value    Income  \
Customer Lifetime Value                       1.000000  0.024366   
Income                                        0.024366  1.000000   
Monthly Premium Auto                          0.396262 -0.016665   
Months Since Last Claim                       0.011517 -0.026715   
Months Since Policy Inception                 0.009418 -0.000875   
Number of Open Complaints                    -0.036343  0.006408   
Number of Policies                            0.021955 -0.008656   
Total Claim Amount                            0.226451 -0.355254   

                               Monthly Premium Auto  Months Since Last Claim  \
Customer Lifetime Value                    0.396262                 0.011517   
Income                                    -0.016665                -0.026715   
Monthly Premium Auto                       1.000000                 0.005026   
Months Since Last Claim                    0.005026                 1.000000   
Months Since Policy Inception              0.020257                -0.042959   
Number of Open Complaints                 -0.013122                 0.005354   
Number of Policies                        -0.011233                 0.009136   
Total Claim Amount                         0.632017                 0.007563   

                               Months Since Policy Inception  \
Customer Lifetime Value                             0.009418   
Income                                             -0.000875   
Monthly Premium Auto                                0.020257   
Months Since Last Claim                            -0.042959   
Months Since Policy Inception                       1.000000   
Number of Open Complaints                          -0.001158   
Number of Policies                                 -0.013333   
Total Claim Amount                                  0.003335   

                               Number of Open Complaints  Number of Policies  \
Customer Lifetime Value                        -0.036343            0.021955   
Income                                          0.006408           -0.008656   
Monthly Premium Auto                           -0.013122           -0.011233   
Months Since Last Claim                         0.005354            0.009136   
Months Since Policy Inception                  -0.001158           -0.013333   
Number of Open Complaints                       1.000000            0.001498   
Number of Policies                              0.001498            1.000000   
Total Claim Amount                             -0.014241           -0.002354   

                               Total Claim Amount  
Customer Lifetime Value                  0.226451  
Income                                  -0.355254  
Monthly Premium Auto                     0.632017  
Months Since Last Claim                  0.007563  
Months Since Policy Inception            0.003335  
Number of Open Complaints               -0.014241  
Number of Policies                      -0.002354  
Total Claim Amount                       1.000000

## Prepare the data for Machine Learning algorithms


In [11]:
df1 = df.copy()

In [12]:
#Remove irrelevant features 
df1 = df1.drop("Customer", axis = 1)
df1 = df1.drop("Effective To Date", axis = 1)

In [14]:
#We might want to remove either policy or policy type (because highly correlated) as we learned 
df1 = df1.drop("Policy Type", axis = 1)

In [5]:
#Do some feature engineering an retrieve the month and day ??

In [15]:
#make sure there is not NA values 
df1 = df1.dropna() #We have no missing value (not imputer needed)

In [16]:
df1.columns

Index(['State', 'Customer Lifetime Value', 'Response', 'Coverage', 'Education',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy', 'Renew Offer Type', 'Sales Channel',
       'Total Claim Amount', 'Vehicle Class', 'Vehicle Size'],
      dtype='object')

In [17]:
#Encode categorical 
binary_feature = df1.nunique()[df1.nunique() == 2].keys().tolist()
categorical_feature = [col for col in df1.select_dtypes('object').columns.to_list() if col not in binary_feature]

In [18]:
# Feature Engineering  - dummyfying the binary and categorical variables
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
for i in binary_feature:
    df1[i] = le.fit_transform(df1[i])

In [19]:

for col in  categorical_feature:
    df1 = pd.concat([df1.drop(col, axis=1),
                            pd.get_dummies(df1[col], prefix=col, prefix_sep='_',
                                           drop_first=True, dummy_na=False)], axis=1)

In [20]:
df1.head()

Customer Lifetime Value  Response  Gender  Income  Monthly Premium Auto  \
0              2763.519279         0       0   56274                    69   
1              6979.535903         0       0       0                    94   
2             12887.431650         0       0   48767                   108   
3              7645.861827         0       1       0                   106   
4              2813.692575         0       1   43836                    73   

   Months Since Last Claim  Months Since Policy Inception  \
0                       32                              5   
1                       13                             42   
2                       18                             38   
3                       18                             65   
4                       12                             44   

   Number of Open Complaints  Number of Policies  Total Claim Amount  ...  \
0                          0                   1          384.811147  ...   
1                          0                   8         1131.464935  ...   
2                          0                   2          566.472247  ...   
3                          0                   7          529.881344  ...   
4                          0                   1          138.130879  ...   

   Sales Channel_Branch  Sales Channel_Call Center  Sales Channel_Web  \
0                     0                          0                  0   
1                     0                          0                  0   
2                     0                          0                  0   
3                     0                          1                  0   
4                     0                          0                  0   

   Vehicle Class_Luxury Car  Vehicle Class_Luxury SUV  Vehicle Class_SUV  \
0                         0                         0                  0   
1                         0                         0                  0   
2                         0                         0                  0   
3                         0                         0                  1   
4                         0                         0                  0   

   Vehicle Class_Sports Car  Vehicle Class_Two-Door Car  Vehicle Size_Medsize  \
0                         0                           1                     1   
1                         0                           0                     1   
2                         0                           1                     1   
3                         0                           0                     1   
4                         0                           0                     1   

   Vehicle Size_Small  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 49 columns]

## Splitting the data : train and test sets 

In [21]:
X = df1.drop(columns=["Customer Lifetime Value"])
y = df1["Customer Lifetime Value"]

In [22]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [23]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_scaled ,y,test_size=0.30,random_state=0)

In [24]:
X_train = pd.DataFrame(X_train,columns = X.columns)
X_test = pd.DataFrame(X_test,columns = X.columns)

## Feature Selection

In [25]:
#Recursive Feature Elimination Method
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(random_state=0)
rfe = RFE(rf, n_features_to_select=25)
model_l = rfe.fit(X_train, y_train)
model_l_df = pd.DataFrame(list(zip(X.columns,model_l.ranking_)), columns = ['predictor','ranking'])
model_l_df

predictor  ranking
0                         Response        1
1                           Gender        1
2                           Income        1
3             Monthly Premium Auto        1
4          Months Since Last Claim        1
5    Months Since Policy Inception        1
6        Number of Open Complaints        1
7               Number of Policies        1
8               Total Claim Amount        1
9                 State_California        1
10                    State_Nevada        6
11                    State_Oregon        2
12                State_Washington        1
13               Coverage_Extended        8
14                Coverage_Premium        1
15               Education_College        1
16                Education_Doctor       16
17  Education_High School or Below        1
18                Education_Master       13
19       EmploymentStatus_Employed        4
20  EmploymentStatus_Medical Leave        9
21        EmploymentStatus_Retired       22
22     EmploymentStatus_Unemployed       18
23          Location Code_Suburban       17
24             Location Code_Urban        1
25          Marital Status_Married        1
26           Marital Status_Single        5
27             Policy_Corporate L2       20
28             Policy_Corporate L3       14
29              Policy_Personal L1       12
30              Policy_Personal L2        1
31              Policy_Personal L3        1
32               Policy_Special L1       24
33               Policy_Special L2       21
34               Policy_Special L3        7
35         Renew Offer Type_Offer2        1
36         Renew Offer Type_Offer3        3
37         Renew Offer Type_Offer4        1
38            Sales Channel_Branch        1
39       Sales Channel_Call Center        1
40               Sales Channel_Web        1
41        Vehicle Class_Luxury Car       23
42        Vehicle Class_Luxury SUV       10
43               Vehicle Class_SUV       15
44        Vehicle Class_Sports Car       19
45      Vehicle Class_Two-Door Car       11
46            Vehicle Size_Medsize        1
47              Vehicle Size_Small        1

In [27]:
#Removing the features that are useless to our model

X_train = X_train.drop(columns = ['Vehicle Class_Two-Door Car','Vehicle Class_Sports Car','Vehicle Class_SUV','Vehicle Class_Luxury SUV','Vehicle Class_Luxury Car','Renew Offer Type_Offer3','Policy_Special L3','Policy_Special L2','Policy_Special L1','Policy_Personal L1','Policy_Corporate L3','Policy_Corporate L2','Marital Status_Single','Location Code_Suburban','EmploymentStatus_Unemployed','EmploymentStatus_Retired','EmploymentStatus_Medical Leave','EmploymentStatus_Employed','Education_Master','Education_Doctor','Coverage_Extended','State_Oregon','State_Nevada'])
X_test = X_test.drop(columns = ['Vehicle Class_Two-Door Car','Vehicle Class_Sports Car','Vehicle Class_SUV','Vehicle Class_Luxury SUV','Vehicle Class_Luxury Car','Renew Offer Type_Offer3','Policy_Special L3','Policy_Special L2','Policy_Special L1','Policy_Personal L1','Policy_Corporate L3','Policy_Corporate L2','Marital Status_Single','Location Code_Suburban','EmploymentStatus_Unemployed','EmploymentStatus_Retired','EmploymentStatus_Medical Leave','EmploymentStatus_Employed','Education_Master','Education_Doctor','Coverage_Extended','State_Oregon','State_Nevada'])

## Train model : Simple Linear Regression

In [28]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [29]:
lin_reg.coef_

array([-144.10260583,  -43.29392813,  149.26310029, 2946.66114005,
         54.44708413,  -26.73253529, -221.42143615,  116.12009075,
        -82.83122563,   66.84115381,   38.65051876, -165.82159887,
         17.19628684,  190.04998799,  167.48274775,  135.57710776,
         62.22247469,   41.60089927, -279.59188759, -236.18304926,
         44.30967693,   58.75246758,  -36.83554429,   96.37922851,
        157.03890097])

### RMSE of training set - LR

In [30]:
from sklearn.metrics import mean_squared_error
import numpy as np 

clv_predictions = lin_reg.predict(X_train)
lin_mse = mean_squared_error(y_train, clv_predictions)
lin_rmse = np.sqrt(lin_mse)
lin_rmse

6335.416848080528

## Train model : Decision Tree

In [31]:
from sklearn.tree import DecisionTreeRegressor

tree_reg = DecisionTreeRegressor(random_state=42)
tree_reg.fit(X_train, y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=42, splitter='best')

### RMSE train set - DT

In [32]:
clv_predictions = tree_reg.predict(X_train)
tree_mse = mean_squared_error(y_train, clv_predictions)
tree_rmse = np.sqrt(tree_mse)
tree_rmse

2.0506403645151865e-13

It is extremely small...

In [33]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(tree_reg, X_train, y_train,
                         scoring="neg_mean_squared_error", cv=10)
tree_rmse_scores = np.sqrt(-scores)


In [34]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation:", scores.std())
display_scores(tree_rmse_scores)

Scores: [6157.41558778 5801.47806111 5048.47065651 5790.84098587 5099.29648932
 5477.71227214 4954.84149196 4446.06984967 5760.1653984  4856.7980166 ]
Mean: 5339.30888093612
Standard deviation: 510.27018127002697


When checking the RMSE of the Decision tree, it is suprisingly low, which might be a sign of overfitting. However when using cross-validation, the Decision Tree doesn’t look as good as it did earlier. It still performs better  than the Linear Regression as we can see computing those same validation scores for the Linear Regression model :

In [35]:
lin_scores = cross_val_score(lin_reg, X_train, y_train,
                             scoring="neg_mean_squared_error", cv=10)
lin_rmse_scores = np.sqrt(-lin_scores)
display_scores(lin_rmse_scores)

Scores: [6803.07254099 6806.27526834 5545.52476959 6947.10197351 6155.43554565
 6418.79042905 6057.15398173 5814.63264252 6474.8467106  6494.91291639]
Mean: 6351.774677836481
Standard deviation: 432.28375155368855


## Train Model: RandomForestRegressor.

Let's try Random Forests work by training many Decision Trees on random subsets of the features, then averaging out their predictions. Building a model on top of many other models is called Ensemble Learning, and it is often a great way to push ML algorithms even further. 

In [52]:
from sklearn.ensemble import RandomForestRegressor

forest_reg = RandomForestRegressor(n_estimators=100, random_state=0)
forest_reg.fit(X_train, y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=0, verbose=0, warm_start=False)

In [53]:
clv_predictions_RF = forest_reg.predict(X_train)
RF_mse = mean_squared_error(y_train, clv_predictions_RF)
RF_rmse = np.sqrt(RF_mse)
RF_rmse

1437.4509527929397

In [55]:
rf_scores = cross_val_score(forest_reg, X_train, y_train,
                             scoring="neg_mean_squared_error", cv=20)
rf_rmse_scores = np.sqrt(-rf_scores)
display_scores(rf_rmse_scores)

Scores: [4909.56470057 3677.62215898 4184.08215711 3971.73822451 3316.70428696
 3967.0261507  5085.59078701 4137.14801191 3911.54655139 3102.89244431
 3958.86764445 3993.4407657  3851.87482296 3098.07106082 3623.57640757
 3279.47988801 3743.26004241 3645.529496   4373.40466705 3446.91426112]
Mean: 3863.9167264772273
Standard deviation: 510.3119540217236


From the Cross-validation results, we can see that the Random Forest performs better than the Decision Tree and the Linear Regression. 

## Train Model: SVM

In [39]:
from sklearn.svm import SVR

svm_reg = SVR(kernel="linear")
svm_reg.fit(X_train, y_train)
clv_predictions_SVM = svm_reg.predict(X_train)
svm_mse = mean_squared_error(y_train, clv_predictions_SVM)
svm_rmse = np.sqrt(svm_mse)
svm_rmse

6954.828562576236

In [40]:
svm_scores = cross_val_score(svm_reg, X_train, y_train,
                             scoring="neg_mean_squared_error", cv=10)
svm_rmse_scores = np.sqrt(-svm_scores)
display_scores(svm_rmse_scores)

Scores: [7582.26097755 7528.30844158 5892.86363446 7748.54500401 6789.04912084
 7072.85055159 6384.74347101 6120.42303399 7032.17642375 7299.74072835]
Mean: 6945.096138711447
Standard deviation: 605.5532519013573


The SVM performs quite poorly. We can therefore infer taht our best model seems to be the Random Forest; we now will be fine-tuning the model with the Hyperparameters.

# Fine-Tune The Model

Now, we have a shortlist of promising models. We now need to fine-tune them. 

In [59]:
##If it is not linear reg or svm (thus, a model with hyperparameters): 
from sklearn.model_selection import GridSearchCV

param_grid = [
    # try 12 (3×4) combinations of hyperparameters
    {'n_estimators': [3, 10, 30], 'max_features': [2, 4, 6, 8]},
    # then try 6 (2×3) combinations with bootstrap set as False
    {'bootstrap': [False], 'n_estimators': [3, 10], 'max_features': [2, 3, 4]},
  ]

forest_reg = RandomForestRegressor(random_state=0)
# train across 5 folds, that's a total of (12+6)*5=90 rounds of training 
grid_search = GridSearchCV(forest_reg, param_grid, cv=20,
                           scoring='neg_mean_squared_error',
                           return_train_score=True)
grid_search.fit(X_train, y_train)
grid_search.best_estimator_



RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features=8, max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=30, n_jobs=None, oob_score=False,
                      random_state=0, verbose=0, warm_start=False)

In [60]:
grid_search.best_params_

{'max_features': 8, 'n_estimators': 30}

In [61]:
cvres = grid_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

6117.971137712389 {'max_features': 2, 'n_estimators': 3}
5301.408761797615 {'max_features': 2, 'n_estimators': 10}
4988.043384306854 {'max_features': 2, 'n_estimators': 30}
5026.346989480905 {'max_features': 4, 'n_estimators': 3}
4475.7355196600065 {'max_features': 4, 'n_estimators': 10}
4340.351888539458 {'max_features': 4, 'n_estimators': 30}
4842.924898428411 {'max_features': 6, 'n_estimators': 3}
4238.446457685696 {'max_features': 6, 'n_estimators': 10}
4030.0629165190385 {'max_features': 6, 'n_estimators': 30}
4412.323686729215 {'max_features': 8, 'n_estimators': 3}
4015.900314204761 {'max_features': 8, 'n_estimators': 10}
3868.3139153551156 {'max_features': 8, 'n_estimators': 30}
6054.48606475124 {'bootstrap': False, 'max_features': 2, 'n_estimators': 3}
5221.301527772549 {'bootstrap': False, 'max_features': 2, 'n_estimators': 10}
5433.5993419101 {'bootstrap': False, 'max_features': 3, 'n_estimators': 3}
4853.487826790143 {'bootstrap': False, 'max_features': 3, 'n_estimators': 10

In this example, we obtain the best solution by setting the max_features hyperparameter to 8 and the n_estimators hyperparameter to 30. The RMSE score for this combination is 3868.



## Evaluate our system on the Test Set 

In [62]:
final_model = grid_search.best_estimator_
final_predictions = final_model.predict(X_test)
final_mse = mean_squared_error(y_test, final_predictions)
final_rmse = np.sqrt(final_mse) 
final_rmse

3468.117328671939